### PR TITLE
Docs: Polish profile documentation

### PR DESCRIPTION
**Pull request by gaogaotiantian**
There are a couple of issues with the current (outdated) documentation

* `To profile a function that takes a single argument` is simply misleading. It does not have to be a function, and `a single argument` is .. just wrong.
* Users do not like "code as string" and we should at least provide the way to do it with actual code in the "instant manual".
* The original statement of cprofile not being able to handle `sys.exit()` is wrong - it can. However, #104676 mentioned that it can't handle `SIGTERM`. I moved this statement up - closer to where it should be seen.
* In the limitation section, the claim about clock precision is super old - we know it was there at least 16 years ago, when it was copied from somewhere else. The resolution of the system clock *now* is at ns level. I removed the section directly, we can redo the statement to talk about this - it's still possible that the system clock resolution is an issue, but it's probably much better and definitely not at 0.001s level.

----
:books: Documentation preview :books:: https://cpython-previews--110494.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/profile.rst [hunk 1]
```diff
@@ -121,6 +121,14 @@ results to a file by specifying a filename to the :func:`run` function::
 The :class:`pstats.Stats` class reads profile results from a file and formats
 them in various ways.
 
+You can also do inline profiling with :class:`Profile`::
```
Comment (skirpichev) [inline Doc/library/profile.rst]:
```suggestion
You can also do inline profiling with :class:`~cProfile.Profile`::
```

File: Doc/library/profile.rst [hunk 2]
```diff
@@ -141,6 +149,9 @@ the output by. This only applies when ``-o`` is not supplied.
    .. versionadded:: 3.8
       Added the ``-m`` option to :mod:`profile`.
 
+Note that if the process is terminated by ``SIGTERM`` during profiling, no
```
Comment (skirpichev) [inline Doc/library/profile.rst]: How about SIGKILL? ;)

File: Doc/library/profile.rst [hunk 3]
```diff
@@ -54,7 +54,7 @@ This section is provided for users that "don't want to read the manual." It
 provides a very brief overview, and allows a user to rapidly perform profiling
 on an existing application.
 
-To profile a function that takes a single argument, you can do::
+To profile a statement with the namespace of ``__main__``, you can do::
```
Comment (skirpichev) [inline Doc/library/profile.rst]:
```suggestion
To profile a statement in the current namespace, you can do::
```

File: Doc/library/profile.rst (hunk 1 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst]: Using `cProfile.Profile` makes sense, because the context manager only works with `cProfile`. However, I don't think the `~` should be used.

File: Doc/library/profile.rst (hunk 3 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst]: Actually, you can't :). `run` will literally use the `__main__` namespace, not the local one.

File: Doc/library/profile.rst (hunk 2 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst]: Good point, I changed the description a bit as there might be other cases when the profiling won't work.

File: Doc/library/profile.rst [hunk 4]
```diff
@@ -121,6 +121,14 @@ results to a file by specifying a filename to the :func:`run` function::
 The :class:`pstats.Stats` class reads profile results from a file and formats
 them in various ways.
 
+You can also do inline profiling with :class:`cProfile.Profile`::
```
Comment (skirpichev) [inline Doc/library/profile.rst]: Hmm, there is still a broken link.

On another hand, this example seems to be redundant: there is such one, right in the description of the Profile class.

File: Doc/library/profile.rst (hunk 3 above)
Comment (skirpichev) [reply Doc/library/profile.rst]: Indeed.
```suggestion
To profile a statement with the namespace of :mod:`__main__`, you can do::
```
Or "in the top-level code environment".

Comment (gaogaotiantian) [reply Doc/library/profile.rst]: I don't think "top-level code enviroment" is clearer than "__main__", but using the module might help.

File: Doc/library/profile.rst (hunk 4 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst]: The reason this example is added is that I believe this example should be in the "instant manual" - which the users who are not interested in the details of cProfile/profile can refer to when they just want a quick profile. I'd say, in many (if not most) cases, users do not want to convert their code into a string - so this kind of inline profiling is what they need.

Yes there's the broken link, but that's how the other part of the docs use it ... The problem is this document is technically ``profile``, but we also put ``cProfile`` stuff in it. The context usage is only available for `cProfile` so `cProfile` should be explicit. We do either this, like the other part of the documentation, or a raw code.

**Comment by gaogaotiantian** [comment]
@CAM-Gerlach could you maybe take a look at it? Thanks!

**Comment by gaogaotiantian** [comment]
@AA-Turner do you know who should I reach out to for the docs change? Thanks!

File: Doc/library/profile.rst [hunk 5]
```diff
@@ -54,7 +54,7 @@ This section is provided for users that "don't want to read the manual." It
 provides a very brief overview, and allows a user to rapidly perform profiling
 on an existing application.
 
-To profile a function that takes a single argument, you can do::
+To profile a statement with the namespace of :mod:`__main__`, you can do::
```
Comment (brandtbucher) [inline Doc/library/profile.rst]: I think this phrasing is a bit awkward. Maybe:

```suggestion
To profile a single statement, you can do::
```

I feel like the comment about the `__main__` namespace will probably just confuse most people, right? If you disagree, maybe rephrase as:

```suggestion
To execute and profile a statement in the :mod:`__main__` namespace, you can do::
```

File: Doc/library/profile.rst (hunk 4 above)
Comment (brandtbucher) [reply Doc/library/profile.rst]: Does `` :class:`cProfile.Profile <profile.Profile>` `` work?

File: Doc/library/profile.rst [hunk 6]
```diff
@@ -577,23 +584,20 @@ implementations.
 Limitations
 ===========
 
-One limitation has to do with accuracy of timing information. There is a
-fundamental problem with deterministic profilers involving accuracy.  The most
-obvious restriction is that the underlying "clock" is only ticking at a rate
-(typically) of about .001 seconds.  Hence no measurements will be more accurate
-than the underlying clock.  If enough measurements are taken, then the "error"
-will tend to average out. Unfortunately, removing this first error induces a
-second source of error.
```
Comment (brandtbucher) [inline Doc/library/profile.rst:586]: Personally, I like this paragraph, and it still contains meaningful information. I'd just update it to leave out specifics about resolution.

File: Doc/library/profile.rst [hunk 7]
```diff
@@ -577,23 +584,20 @@ implementations.
...
+To avoid infinite recursion, :mod:`profile` and :mod:`cProfile` do not work
+inside the trace function set by :func:`sys.settrace` or :func:`sys.setprofile`.
```
Comment (brandtbucher) [inline Doc/library/profile.rst]:
```suggestion
inside trace functions set using :func:`sys.settrace` or :func:`sys.setprofile`.
```

Comment (brandtbucher) [reply Doc/library/profile.rst]: I'd also move this down near the bottom, since it seems like a pretty niche issue.

File: Doc/library/profile.rst [hunk 8]
```diff
@@ -577,23 +584,20 @@ implementations.
...
-The second problem is that it "takes a while" from when an event is dispatched
+It "takes a while" from when an event is dispatched
 until the profiler's call to get the time actually *gets* the state of the
 clock.  Similarly, there is a certain lag when exiting the profiler event
 handler from the time that the clock's value was obtained (and then squirreled
 away), until the user's code is once again executing.  As a result, functions
 that are called many times, or call many functions, will typically accumulate
-this error. The error that accumulates in this fashion is typically less than
-the accuracy of the clock (less than one clock tick), but it *can* accumulate
-and become very significant.
```
Comment (brandtbucher) [inline Doc/library/profile.rst:596]: I'd leave this?

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks for your patience.

I'd prefer if we just removed clearly incorrect bits and added additional info, rather than reworking the "Limitations" section this extensively. Let me know what you think.

File: Doc/library/profile.rst (hunk 5 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst]: Yeah the comment about `__main__` will probably confuse users as it's a relatively deep detail. Especially considering this is the "instant user manual". So I'll take that out.

File: Doc/library/profile.rst (hunk 4 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst]: Yes this works. Nice trick!

File: Doc/library/profile.rst (hunk 6 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst:586]: Sure. But in reality, the clock resolution should not be an issue (or significant issue compared to other source) anymore. If we list this as limitation, there are a LOT of factors that we could potentially write here.

File: Doc/library/profile.rst (hunk 8 above)
Comment (gaogaotiantian) [reply Doc/library/profile.rst:596]: The sentence `the error that accumulates in this fashion is typically less than
the accuracy of the clock (less than one clock tick)` is incorrect now, as the accuracy of the clock has improved a lot for the last .. I don't know how many years. This is definitely more significant than the clock accuracy issue. Do you want me to keep the "become very significant"?
